libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1947,86 +2014,174 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
+				{
+					duplication = true;
+					m_errorReporter.typeError(
+						arguments[i]->location(),
+						"Duplicate named argument."
+					);
+				}
+
+		// check actual types
+		if (!duplication)
 			for (size_t i = 0; i < argumentNames.size(); i++)
-				for (size_t j = i + 1; j < argumentNames.size(); j++)
-					if (*argumentNames[i] == *argumentNames[j])
+			{
+				bool found = false;
+				for (size_t j = 0; j < parameterNames.size(); j++)
+					if (parameterNames[j] == *argumentNames[i])
 					{
-						duplication = true;
-						m_errorReporter.typeError(arguments[i]->location(), "Duplicate named argument.");
+						found = true;
+						// check type convertible
+						if (!type(*arguments[i])->isImplicitlyConvertibleTo(*parameterTypes[j]))
+							m_errorReporter.typeError(
+								arguments[i]->location(),
+								"Invalid type for argument in function call. "
+								"Invalid implicit conversion from " +
+								type(*arguments[i])->toString() +
+								" to " +
+								parameterTypes[i]->toString() +
+								" requested."
+							);
+						break;
 					}
 
-			// check actual types
-			if (!duplication)
-				for (size_t i = 0; i < argumentNames.size(); i++)
-				{
-					bool found = false;
-					for (size_t j = 0; j < parameterNames.size(); j++)
-						if (parameterNames[j] == *argumentNames[i])
-						{
-							found = true;
-							// check type convertible
-							if (!type(*arguments[i])->isImplicitlyConvertibleTo(*parameterTypes[j]))
-								m_errorReporter.typeError(
-									arguments[i]->location(),
-									"Invalid type for argument in function call. "
-									"Invalid implicit conversion from " +
-									type(*arguments[i])->toString() +
-									" to " +
-									parameterTypes[i]->toString() +
-									" requested."
-								);
-							break;
-						}
-
-					if (!found)
-						m_errorReporter.typeError(
-							_functionCall.location(),
-							"Named argument \"" + *argumentNames[i] +  "\" does not match function declaration."
-						);
-				}
-		}
+				if (!found)
+					m_errorReporter.typeError(
+						_functionCall.location(),
+						"Named argument \"" +
+						*argumentNames[i] +
+						"\" does not match function declaration."
+					);
+			}
+
+		return;
 	}
+}
 
-	if (returnTypes.size() == 1)
-		_functionCall.annotation().type = returnTypes.front();
+bool TypeChecker::visit(FunctionCall const& _functionCall)
+{
+	vector<ASTPointer<Expression const>> const& arguments = _functionCall.arguments();
+	bool argumentsArePure = true;
+
+	// We need to check arguments' type first as they will be needed for overload resolution.
+	for (ASTPointer<Expression const> const& argument: arguments)
+	{
+		argument->accept(*this);
+		if (!argument->annotation().isPure)
+			argumentsArePure = false;
+	}
+
+	// For positional calls only, store argument types
+	if (_functionCall.names().empty())
+	{
+		shared_ptr<TypePointers> const argumentTypes = make_shared<TypePointers>();
```

Comment:
Can you move from it properly if it is const?

Reply:
Good catch, it's actually the shared_ptr copy ctor that gets called rather than the move ctor which is why this compiled cleanly.